File: food-delivery-app-main/backend-dotnet/FoodDelivery.API/Controllers/UserController.cs
using FoodDelivery.API.DTOs;
using FoodDelivery.API.Services;
using Microsoft.AspNetCore.Mvc;

namespace FoodDelivery.API.Controllers
{
    [ApiController]
    [Route("api/user")]
    public class UserController : ControllerBase
    {
        private readonly UserService _userService;

        public UserController(UserService userService)
        {
            _userService = userService;
        }

        [HttpPost("login")]
        public async Task<IActionResult> Login(LoginRequest request)
        {
            var token = await _userService.Login(request.Email, request.Password);
            if (token == null)
                return Ok(new { success = false, message = "Invalid credentials" });

            return Ok(new { success = true, token });
        }

        [HttpPost("register")]
        public async Task<IActionResult> Register(RegisterRequest request)
        {
            var token = await _userService.Register(
                request.Name, request.Email, request.Password
            );

            if (token == null)
                return Ok(new { success = false, message = "User already exists" });

            return Ok(new { success = true, token });
        }

        [HttpPost("google")]
        public async Task<IActionResult> GoogleLogin(GoogleLoginRequest request)
        {
            var token = await _userService.GoogleLogin(
                request.Name, request.Email, request.Image
            );

            return Ok(new { success = true, token });
        }
    }
}

File: food-delivery-app-main/backend/controllers/userController.js
import userModel from "../models/userModel.js";
import jwt from "jsonwebtoken"
import bcrypt from 'bcrypt'
import validator from 'validator'

//login user
const loginUser = async (req,res) => {
    const {email,password} = req.body;
    try {
        const user = await userModel.findOne({email})

        if (!user) {
            return res.json({success:false, message:"User doesn't exists"})
        }

        const isMatch = await bcrypt.compare(password, user.password)

        if (!isMatch) {
            return res.json({success:false, message: "Invalid credentials"})
        }

        const token = createToken(user._id)
        res.json({success:true, token })

    } catch (error) {
        console.log(error);
        res.json({success:false, message: "Error"})
    }
}

const createToken = (id) => {
    return jwt.sign({id},process.env.JWT_SECRET)
}

//register user
const registerUser = async (req,res) => {
    const {name, password, email} = req.body;
    try {
        // checking user already exists
        const exists = await userModel.findOne({email});
        if (exists) {
            return res.json({success:false,message:"User already exists"})
        }

        //validating email format and strong password
        if (!validator.isEmail(email)) {
            return res.json({success:false,message:"Please enter a valid email"})
        }

        if (password.length<8) {
            return res.json({success:false,message:"Please enter a strong password"})
        }

        //hashing user password
        const salt = await bcrypt.genSalt(10); // can be set from 5 to 15, if it is higher it takes time to encrypt
        const hashedPassword = await bcrypt.hash(password,salt);

        const newUser = new userModel({
            name: name,
            email: email,
            password: hashedPassword
        })

        const user = await newUser.save()
        const token = createToken(user._id)
        res.json({success:true, token})

    } catch (error) {
        console.log(error);
        res.json({success: false, message:"Error"});
    }
}

const googleLogin = async (req, res) => {
    const { name, email, image } = req.body;

    try {
        let user = await userModel.findOne({ email });

        if (!user) {
            user = new userModel({
                name,
                email,
                image,
                cartData: {}
            });
            await user.save();
        }

        const token = createToken(user._id);
        res.json({ success: true, token });

    } catch (error) {
        console.log(error);
        res.json({ success: false, message: "Google login error" });
    }
};


export {loginUser, registerUser, googleLogin}